File: Python3/aoc2020/solutions/day_16.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Dict, Iterator, List, Tuple

from .solution import Solution

FieldName = str
FieldValidator = Callable[[int], bool]
FieldRules = Dict[FieldName, FieldValidator]
Ticket = List[int]


class Graph:
    def __init__(self, n):
        self.g = [[] for _ in range(n)]

    def add_edge(self, u, v):
        self.g[u].append(v)
        self.g[v].append(u)

    def find_matching_using_Kuhn_alogrithm(self):
        match = [-1] * len(self.g)
        visited = set()
        has_augmented = True

        while has_augmented:
            has_augmented = False
            visited.clear()

            for u in range(len(self.g)):
                if match[u] == -1:
                    has_augmented = has_augmented or self.dfs(
                        u, visited, match)

        return match

    def dfs(self, u, visited, match):
        if u in visited:
            return False

        visited.add(u)

        for v in self.g[u]:
            if match[v] == -1 or self.dfs(match[v], visited, match):
                match[v] = u
                match[u] = v
                return True

        return False


@dataclass
class Notes:
    rules: FieldRules
    my_ticket: Ticket
    nearby_tickets: List[Ticket]

    def sum_of_invalid_values_in_nearby_tickets(self) -> int:
        return sum(self.identify_invalid_values_in_nearby_tickets())

    def identify_invalid_values_in_nearby_tickets(self) -> Iterator[int]:
        for nearby_ticket in self.nearby_tickets:
            for number in nearby_ticket:
                if self.is_value_not_valid_for_any_field(number):
                    yield number

    def is_value_not_valid_for_any_field(self, value: int) -> bool:
        return all(not validator(value) for validator in self.rules.values())

    def product_of_departure_values(self) -> int:
        product = 1
        for field, value in zip(self.find_field_order(), self.my_ticket):
            if field.startswith("departure"):
                product *= value

        return product

    def find_field_order(self) -> List[FieldName]:
        valid_tickets = self.find_valid_tickets(
            self.nearby_tickets + [self.my_ticket])

        rules_list = list(self.rules.items())
        fields_count = len(self.rules)
        graph = Graph(2 * fields_count)

        for column in range(fields_count):
            for rule_idx, (name, validator) in enumerate(rules_list, start=fields_count):
                for ticket in valid_tickets:
                    if not validator(ticket[column]):
                        break
                else:
                    graph.add_edge(column, rule_idx)

        matching = graph.find_matching_using_Kuhn_alogrithm()

        return [rules_list[matching[rule_idx]-fields_count][0] for rule_idx in range(fields_count)]

    def find_valid_tickets(self, tickets: List[Ticket]) -> List[Ticket]:
        return [t for t in tickets if self.is_ticket_valid(t)]

    def is_ticket_valid(self, ticket: Ticket) -> bool:
        return not any(self.is_value_not_valid_for_any_field(v) for v in ticket)

    @classmethod
    def from_str(cls, notes_text: str) -> Notes:
        parts = notes_text.split("\n\n")
        rules = cls.parse_rules(parts[0])
        my_ticket = cls.parse_my_ticket(parts[1])
        nearby_tickets = cls.parse_nearby_tickets(parts[2])

        return cls(rules, my_ticket, nearby_tickets)

    @staticmethod
    def parse_my_ticket(my_ticket_text: str) -> Ticket:
        return Notes.parse_ticket(my_ticket_text.lstrip("your ticket:\n"))

    @staticmethod
    def parse_nearby_tickets(nearby_tickets_text: str) -> List[Ticket]:
        tickets_str = nearby_tickets_text.lstrip("nearby tickets:\n")

        return [Notes.parse_ticket(t) for t in tickets_str.splitlines()]

    @staticmethod
    def parse_ticket(ticket_text: str) -> Ticket:
        return [int(i, base=10) for i in ticket_text.split(",")]

    @staticmethod
    def parse_rules(rules_text: str) -> FieldRules:
        rules = {}
        for rule in rules_text.splitlines():
            name, validator = Notes.parse_rule(rule)
            rules[name] = validator
        return rules

    @staticmethod
    def parse_rule(rule_text: str) -> Tuple[FieldName, FieldValidator]:
        name, values = rule_text.split(": ")
        return name, Notes.parse_value_intervals(values)

    @staticmethod
    def parse_value_intervals(value_intervals_text: str) -> FieldValidator:
        ranges = []
        for value_interval in value_intervals_text.split(" or "):
            from_text, to_text = value_interval.split("-")
            from_int = int(from_text, base=10)
            to_int = int(to_text, base=10)

            ranges.append((from_int, to_int))

        return Notes.create_field_validator(ranges)

    @staticmethod
    def create_field_validator(ranges: List[Tuple[int, int]]) -> FieldValidator:
        def validator(value: int) -> bool:
            return any(f <= value <= t for f, t in ranges)

        return validator


class Day16(Solution):

    def first_task(self, notes_text: str) -> str:
        notes = Notes.from_str(notes_text)

        return str(notes.sum_of_invalid_values_in_nearby_tickets())

    def second_task(self, notes_text: str) -> str:
        notes = Notes.from_str(notes_text)

        return str(notes.product_of_departure_values())

File: Rust/src/solutions/day_16.rs
use super::solution::{Error, Solution};
use itertools::Itertools;
use std::str::FromStr;

type Ticket = Vec<i32>;

struct Rule {
    name: String,
    intervals: Vec<(i32, i32)>,
}

impl FromStr for Rule {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut parts = s.split(": ");
        let name = parts.next().unwrap().to_string();
        let intevals_text = parts.next().unwrap().split(" or ");

        let intervals: Vec<(i32, i32)> = intevals_text
            .map(|el| {
                el.split('-')
                    .map(|el| el.parse::<i32>().unwrap())
                    .collect_tuple()
                    .unwrap()
            })
            .collect();

        Ok(Rule { name, intervals })
    }
}

impl Rule {
    fn apply_to(&self, value: i32) -> bool {
        self.intervals
            .iter()
            .any(|(f, t)| *f <= value && value <= *t)
    }
}

struct Graph {
    g: Vec<Vec<usize>>,
}

impl Graph {
    fn new(size: usize) -> Graph {
        Graph {
            g: vec![vec![]; size],
        }
    }

    fn add_edge(&mut self, u: usize, v: usize) {
        self.g[u].push(v);
        self.g[v].push(u);
    }

    fn find_matching_using_kuhn_algorithm(&self) -> Vec<usize> {
        let mut matching = vec![usize::MAX; self.g.len()];
        let mut visited = vec![false; self.g.len()];
        let mut has_augmented = true;

        while has_augmented {
            has_augmented = false;
            for v in &mut visited {
                *v = false;
            }

            for u in 0..self.g.len() {
                if matching[u] == usize::MAX && !visited[u] {
                    has_augmented |= self.dfs(u, &mut visited, &mut matching);
                }
            }
        }

        matching
    }

    fn dfs(&self, u: usize, visited: &mut Vec<bool>, matching: &mut Vec<usize>) -> bool {
        if visited[u] {
            return false;
        }

        visited[u] = true;
        for &v in &self.g[u] {
            if matching[v] == usize::MAX || self.dfs(matching[v], visited, matching) {
                matching[v] = u;
                matching[u] = v;
                return true;
            }
        }

        false
    }
}

struct Notes {
    rules: Vec<Rule>,
    my_ticket: Ticket,
    nearby_tickets: Vec<Ticket>,
}

impl Notes {
    fn find_error_rate(&self) -> i64 {
        self.nearby_tickets
            .iter()
            .flat_map(|ticket| ticket.iter())
            .filter(|&&num| self.rules.iter().all(|r| !r.apply_to(num)))
            .map(|&el| el as i64)
            .sum()
    }

    fn find_product_of_departure_fields(&self) -> i64 {
        self.find_field_order()
            .iter()
            .zip(self.my_ticket.iter())
            .filter(|(name, _)| name.starts_with("departure"))
            .map(|(_, &v)| v as i64)
            .product()
    }

    fn find_field_order(&self) -> Vec<&String> {
        let valid_tickets = self.find_valid_tickets();
        let rules_count = self.rules.len();
        let mut graph = Graph::new(2 * rules_count);

        for column in 0..self.my_ticket.len() {
            for (rule_idx, rule) in self.rules.iter().enumerate() {
                let mut acceptable_field = true;
                for ticket in valid_tickets.iter() {
                    if !rule.apply_to(ticket[column]) {
                        acceptable_field = false;
                        break;
                    }
                }

                if acceptable_field {
                    graph.add_edge(column, rule_idx + rules_count);
                }
            }
        }

        let matching = graph.find_matching_using_kuhn_algorithm();

        (0..rules_count)
            .map(|idx| &self.rules[matching[idx] - rules_count].name)
            .collect()
    }

    fn find_valid_tickets(&self) -> Vec<&Ticket> {
        let mut valid_tickets: Vec<&Ticket> = self
            .nearby_tickets
            .iter()
            .filter(|ticket| self.is_ticket_valid(ticket))
            .collect();

        valid_tickets.push(&self.my_ticket);

        valid_tickets
    }

    fn is_ticket_valid(&self, ticket: &Ticket) -> bool {
        ticket
            .iter()
            .all(|&num| self.rules.iter().any(|r| r.apply_to(num)))
    }

    fn parse_rules(rules_text: &str) -> Vec<Rule> {
        rules_text.lines().map(|el| el.parse().unwrap()).collect()
    }

    fn parse_my_ticket(my_ticket_text: &str) -> Ticket {
        Notes::parse_ticket(my_ticket_text.trim_start_matches("your ticket:\n"))
    }

    fn parse_nearby_tickets(nearby_tickets_text: &str) -> Vec<Ticket> {
        nearby_tickets_text
            .trim_start_matches("nearby tickets:\n")
            .lines()
            .map(Notes::parse_ticket)
            .collect()
    }

    fn parse_ticket(ticket_text: &str) -> Ticket {
        ticket_text
            .split(',')
            .map(|el| el.parse().unwrap())
            .collect()
    }
}

impl FromStr for Notes {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (rules, my_ticket, nearby_tickets) = s.split("\n\n").collect_tuple().unwrap();
        let rules = Notes::parse_rules(rules);
        let my_ticket = Notes::parse_my_ticket(my_ticket);
        let nearby_tickets = Notes::parse_nearby_tickets(nearby_tickets);

        Ok(Notes {
            rules,
            my_ticket,
            nearby_tickets,
        })
    }
}

pub struct Day16 {}

impl Solution for Day16 {
    fn first_task(&self, notes_text: &str) -> Result<String, Error> {
        let notes: Notes = notes_text.parse().unwrap();
        Ok(notes.find_error_rate().to_string())
    }

    fn second_task(&self, notes_text: &str) -> Result<String, Error> {
        let notes: Notes = notes_text.parse().unwrap();
        Ok(notes.find_product_of_departure_fields().to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example_error_rate() {
        let notes_text = "class: 1-3 or 5-7
row: 6-11 or 33-44
seat: 13-40 or 45-50

your ticket:
7,1,14

nearby tickets:
7,3,47
40,4,50
55,2,20
38,6,12";
        let notes: Notes = notes_text.parse().unwrap();

        assert_eq!(71, notes.find_error_rate());
    }

    #[test]
    fn test_example_row_order() {
        let notes_text = "class: 0-1 or 4-19
row: 0-5 or 8-19
seat: 0-13 or 16-19

your ticket:
11,12,13

nearby tickets:
3,9,18
15,1,5
5,14,9";
        let notes: Notes = notes_text.parse().unwrap();

        assert_eq!(vec!["row", "class", "seat"], notes.find_field_order());
    }
}
